# **Abstract**
## Implementaion of Q Learning Reinforcement Learning to play Frozen Lake
The frozen lake environment is a 4 × 4 grid with four possible zones: Safe (S), Frozen (F), Hole (H), and Target (G). The agent moves around the grid until it reaches the target or hole. If he falls into the hole, he should start again and be rewarded with 0.. This process continues until it learns from each mistake and finally reaches the goal.
The agent in the environment has four possible moves — Up, Down, Left and Right. This environment will allow the agent to move accordingly.This grid has 16 possible blocks where the agent will be at a given time. At the current state, the agent will have four possibilities of deciding the next state.



#### **Installation of Requried Libraries**

In [0]:
!pip install numpy
!pip install gym

#### **Import Requried Libraries**

In [0]:
import numpy as np
import gym
import random

Creation of Frozon Lake enviroment

In [0]:
env = gym.make("FrozenLake-v0")

To determine total number of rows and column of Q table, Action_size and State_size need to be determined

In [0]:
action_size = env.action_space.n
print('Action Size: ', action_size)

Action Size:  4


In [0]:
state_size = env.observation_space.n
print('State Size: ', state_size)

State Size:  16


Initialization of Qtable

In [0]:
def init_qtable():
  qtable = np.zeros((state_size, action_size))
  print(qtable)

Base Line Model Parameters

In [0]:
rewards = []
total_episodes = 5000       
total_test_episodes = 100
max_steps = 99              
learning_rate = 0.7
gamma = 0.8
epsilon = 1.0               
max_epsilon = 1.0           
min_epsilon = 0.1           
decay_rate = 0.01       
qtable = np.zeros((state_size, action_size))


Creation  of Q-table by taking total number of episodes 

In [0]:
for episode in range(total_episodes):
    state = env.reset()
    step = 0
    done = False
    total_rewards = 0   
    for step in range(max_steps):
        exp_exp_tradeoff = random.uniform(0, 1) 
        if exp_exp_tradeoff > epsilon:
            action = np.argmax(qtable[state,:])
        else:
            action = env.action_space.sample()
        new_state, reward, done, info = env.step(action)
        qtable[state, action] = qtable[state, action] + learning_rate * (reward + gamma * np.max(qtable[new_state, :]) - qtable[state, action])
        total_rewards += reward
        state = new_state

        if done == True: 
            break

    epsilon = min_epsilon + (max_epsilon - min_epsilon)*np.exp(-decay_rate*episode) 
    rewards.append(total_rewards)

print ("Score over time: " +  str(sum(rewards)/total_episodes))

Score over time: 0.1022


Calculation of average reward per thousand episodes


In [0]:
  rewards_per_thosand_episodes = np.split(np.array(rewards),total_episodes/1000)
  count = 1000  
  print("Average reward per thousand episodes")
  for r in rewards_per_thosand_episodes:
      print(count, ": ", str(sum(r/1000)))
      count += 1000

Average reward per thousand episodes
1000 :  0.001
2000 :  0.1210000000000001
3000 :  0.12700000000000009
4000 :  0.1330000000000001
5000 :  0.1290000000000001


Display Q table

In [0]:
print(qtable)

[[7.46493488e-03 2.74594444e-02 4.61159617e-03 6.39136989e-03]
 [6.85356938e-03 1.64855394e-04 2.30354866e-03 4.33464352e-02]
 [8.41008512e-02 3.53628886e-03 3.13907264e-03 6.28435557e-04]
 [7.64133942e-05 4.89982432e-04 1.34248833e-03 3.54135940e-03]
 [4.06984672e-02 1.36462558e-02 2.78259153e-03 1.22725937e-03]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00]
 [8.27636732e-03 2.61630379e-01 3.65743709e-06 1.95960324e-04]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00]
 [8.22063980e-03 5.77721799e-02 5.50241660e-02 8.95269776e-03]
 [1.13511758e-01 1.78603289e-01 4.38920787e-02 5.63382058e-02]
 [1.47825334e-01 2.88707059e-02 3.27033212e-02 6.65914433e-03]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00]
 [3.81790301e-02 3.19682089e-02 2.32337225e-01 1.20640044e-02]
 [2.02366620e-01 3.71880694e-01 2.89239863e-01 2.28175626e-01]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 0.000000

Testing baseline model

In [0]:
env.reset()
for episode in range(total_test_episodes):
    state = env.reset()
    step = 0
    done = False
    print("****************************************************")
    print("Episode ", episode)
    for step in range(max_steps):     
        action = np.argmax(qtable[state,:])
        new_state, reward, done, info = env.step(action)        
        if done:
            env.render()
            if new_state == 15:
                print("You Won 🏆")
            else:
                print("Game Over ☠️")
            print("Total number of steps in episode", step)            
            break
        state = new_state
env.close()

****************************************************
Episode  0
  (Down)
SFFF
FHFH
FFFH
HFFG
Game Over ☠️
Total number of steps in episode 4
****************************************************
Episode  1
  (Down)
SFFF
FHFH
FFFH
HFFG
You Won 🏆
Total number of steps in episode 18
****************************************************
Episode  2
  (Down)
SFFF
FHFH
FFFH
HFFG
Game Over ☠️
Total number of steps in episode 4
****************************************************
Episode  3
  (Down)
SFFF
FHFH
FFFH
HFFG
Game Over ☠️
Total number of steps in episode 6
****************************************************
Episode  4
  (Down)
SFFF
FHFH
FFFH
HFFG
Game Over ☠️
Total number of steps in episode 12
****************************************************
Episode  5
  (Down)
SFFF
FHFH
FFFH
HFFG
Game Over ☠️
Total number of steps in episode 11
****************************************************
Episode  6
  (Down)
SFFF
FHFH
FFFH
HFFG
Game Over ☠️
Total number of steps in episode 7
*************

Tuning Hyperparameters

In [0]:
rewards = []
total_episodes = 5000       
total_test_episodes = 100
max_steps = 99              
learning_rate = 0.92
gamma = 0.95
epsilon = 1.0               
max_epsilon = 1.0           
min_epsilon = 0.1           
decay_rate = 0.01 
init_qtable()

[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]


Creation  of Q-table by taking total number of episodes 
  

In [0]:
for episode in range(total_episodes):
    state = env.reset()
    step = 0
    done = False
    total_rewards = 0   
    for step in range(max_steps):
        exp_exp_tradeoff = random.uniform(0, 1) 
        if exp_exp_tradeoff > epsilon:
            action = np.argmax(qtable[state,:])
        else:
            action = env.action_space.sample()
        new_state, reward, done, info = env.step(action)
        qtable[state, action] = qtable[state, action] + learning_rate * (reward + gamma * np.max(qtable[new_state, :]) - qtable[state, action])
        total_rewards += reward
        state = new_state

        if done == True: 
            break

    epsilon = min_epsilon + (max_epsilon - min_epsilon)*np.exp(-decay_rate*episode) 
    rewards.append(total_rewards)

print ("Score over time: " +  str(sum(rewards)/total_episodes))


Score over time: 0.1504



Calculation of average reward per thousand episodes
    

In [0]:
rewards_per_thosand_episodes = np.split(np.array(rewards),total_episodes/1000)
count = 1000  
print("Average reward per thousand episodes")
for r in rewards_per_thosand_episodes:
    print(count, ": ", str(sum(r/1000)))
    count += 1000


Average reward per thousand episodes
1000 :  0.12700000000000009
2000 :  0.16800000000000012
3000 :  0.16300000000000012
4000 :  0.1290000000000001
5000 :  0.16500000000000012


Display Q table


In [0]:
print(qtable)


[[1.81871117e-01 1.74879717e-01 1.73902456e-01 1.68273516e-01]
 [1.64889571e-03 1.89159348e-03 5.94784433e-03 3.91759650e-02]
 [3.92662420e-02 3.15883982e-03 5.93097850e-04 1.39810876e-02]
 [1.81703989e-03 1.41935302e-05 1.07729115e-02 1.11507031e-02]
 [1.99789726e-01 1.77941767e-01 1.91380717e-01 1.82842615e-01]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00]
 [3.32096865e-05 1.69827994e-06 2.25621537e-05 1.53445419e-05]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00]
 [2.28905849e-04 3.03313005e-04 3.37773468e-03 3.89751232e-01]
 [5.29572705e-03 8.18429461e-01 9.35190533e-03 1.26207253e-01]
 [1.13602726e-03 9.45765070e-01 2.94143986e-03 3.76541551e-03]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00]
 [1.46339107e-01 4.20392125e-02 5.05365097e-01 4.12118901e-03]
 [6.41203085e-01 9.99985841e-01 1.73616352e-01 6.21118663e-01]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 0.000000

Testing Tuned model

In [0]:
env.reset()
for episode in range(total_test_episodes):
    state = env.reset()
    step = 0
    done = False
    print("****************************************************")
    print("Episode ", episode)
    total_steps = 0
    total_episode = 0
    for step in range(max_steps):     
        action = np.argmax(qtable[state,:])
        new_state, reward, done, info = env.step(action)        
        if done:
            env.render()
            if new_state == 15:
                print("You Won 🏆")
            else:
                print("Game Over ☠️")
            print("Total number of steps in episode", step)   
            total_steps = total_steps + step   
            total_episode = total_episode + episode    
            break
        state = new_state
env.close()
print('Average steps taken per episode: ', total_steps/total_episode )

****************************************************
Episode  0
  (Down)
SFFF
FHFH
FFFH
HFFG
You Won 🏆
Total number of steps in episode 48
****************************************************
Episode  1
  (Down)
SFFF
FHFH
FFFH
HFFG
Game Over ☠️
Total number of steps in episode 28
****************************************************
Episode  2
  (Down)
SFFF
FHFH
FFFH
HFFG
You Won 🏆
Total number of steps in episode 15
****************************************************
Episode  3
  (Down)
SFFF
FHFH
FFFH
HFFG
You Won 🏆
Total number of steps in episode 17
****************************************************
Episode  4
  (Down)
SFFF
FHFH
FFFH
HFFG
Game Over ☠️
Total number of steps in episode 18
****************************************************
Episode  5
  (Down)
SFFF
FHFH
FFFH
HFFG
Game Over ☠️
Total number of steps in episode 34
****************************************************
Episode  6
  (Down)
SFFF
FHFH
FFFH
HFFG
You Won 🏆
Total number of steps in episode 30
******************

Changing Epsilon and decay_rate in Baseline Model

In [0]:
rewards = []
total_episodes = 5000       
total_test_episodes = 100
max_steps = 99              
learning_rate = 0.7
gamma = 0.8
epsilon = 0.6           
max_epsilon = 0.6           
min_epsilon = 0.1           
decay_rate = 0.04
init_qtable()


[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]


In [0]:
#Creation of Q-table by taking total number of episodes
for episode in range(total_episodes):
    state = env.reset()
    step = 0
    done = False
    total_rewards = 0   
    for step in range(max_steps):
        exp_exp_tradeoff = random.uniform(0, 1) 
        if exp_exp_tradeoff > epsilon:
            action = np.argmax(qtable[state,:])
        else:
            action = env.action_space.sample()
        new_state, reward, done, info = env.step(action)
        qtable[state, action] = qtable[state, action] + learning_rate * (reward + gamma * np.max(qtable[new_state, :]) - qtable[state, action])
        total_rewards += reward
        state = new_state

        if done == True: 
            break

    epsilon = min_epsilon + (max_epsilon - min_epsilon)*np.exp(-decay_rate*episode) 
    rewards.append(total_rewards)

print ("Score over time: " +  str(sum(rewards)/total_episodes))


Score over time: 0.1172


In [0]:
#Calculation of average reward per thousand episodes
rewards_per_thosand_episodes = np.split(np.array(rewards),total_episodes/1000)
count = 1000  
print("Average reward per thousand episodes")
for r in rewards_per_thosand_episodes:
    print(count, ": ", str(sum(r/1000)))
    count += 1000


Average reward per thousand episodes
1000 :  0.12700000000000009
2000 :  0.11300000000000009
3000 :  0.1230000000000001
4000 :  0.09300000000000007
5000 :  0.1300000000000001


In [0]:
#Display Q table
print(qtable)


[[6.40948581e-03 2.58264586e-03 2.68978489e-03 2.56155395e-02]
 [2.21257326e-03 3.79012845e-04 3.68038842e-04 6.15000723e-02]
 [2.88399336e-01 5.53224402e-03 6.21587194e-03 5.79524205e-03]
 [4.55824735e-03 6.73020293e-03 4.80602072e-04 5.47717678e-03]
 [3.33242665e-03 2.29708462e-03 2.07943078e-03 1.81362777e-03]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00]
 [4.65966961e-01 1.21108438e-01 4.25225630e-03 2.60794544e-05]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00]
 [4.04605382e-03 1.31027337e-03 2.76029078e-03 4.18547459e-03]
 [6.77077438e-03 6.60208417e-02 4.38402058e-03 2.58935944e-03]
 [7.85614986e-01 6.16335500e-02 2.16840479e-03 1.58120393e-03]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 0.00000000e+00]
 [9.91973254e-03 9.88261525e-02 8.65555480e-02 6.75632912e-03]
 [2.97829677e-02 9.98194415e-01 1.39500231e-01 1.76133252e-01]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 0.000000

In [0]:
#Testing new baseline model

env.reset()
for episode in range(total_test_episodes):
    state = env.reset()
    step = 0
    done = False
    print("****************************************************")
    print("Episode ", episode)
    total_steps = 0
    total_episode = 0
    for step in range(max_steps):     
        action = np.argmax(qtable[state,:])
        new_state, reward, done, info = env.step(action)        
        if done:
            env.render()
            if new_state == 15:
                print("You Won 🏆")
            else:
                print("Game Over ☠️")
            print("Total number of steps in episode", step)   
            total_steps = total_steps + step   
            total_episode = total_episode + episode    
            break
        state = new_state
env.close()
print('Average steps taken per episode: ', total_steps/total_episode )

****************************************************
Episode  0
  (Left)
SFFF
FHFH
FFFH
HFFG
Game Over ☠️
Total number of steps in episode 41
****************************************************
Episode  1
  (Left)
SFFF
FHFH
FFFH
HFFG
Game Over ☠️
Total number of steps in episode 7
****************************************************
Episode  2
  (Down)
SFFF
FHFH
FFFH
HFFG
You Won 🏆
Total number of steps in episode 13
****************************************************
Episode  3
  (Down)
SFFF
FHFH
FFFH
HFFG
Game Over ☠️
Total number of steps in episode 26
****************************************************
Episode  4
  (Left)
SFFF
FHFH
FFFH
HFFG
Game Over ☠️
Total number of steps in episode 6
****************************************************
Episode  5
  (Left)
SFFF
FHFH
FFFH
HFFG
Game Over ☠️
Total number of steps in episode 18
****************************************************
Episode  6
  (Down)
SFFF
FHFH
FFFH
HFFG
Game Over ☠️
Total number of steps in episode 24
***********

**Establish a baseline performance. How well did your RL Q-learning do on your problem?**

Below are the observations when Q-learning model implemented using baseline paramenters:

*Score over time:* 0.1158

*Average reward per thousand episodes*

1000 :  0.09500000000000007

2000 :  0.1460000000000001 

3000 :  0.11300000000000009 

4000 :  0.09800000000000007 

5000 :  0.12700000000000009 


**What are the states, the actions and the size of the Q-table?**

Action Size:  4(Starting State (S), Frozen Tiles (F), Holes (H) and the Goal state (G))

State Size:  16

Qtable Size: 16 * 4

**What are the rewards? Why did you choose them?**

Particular episode will end when agent reach at goal or falls in the hole. Agent will receive a reward 1 if it reach the goal otherwise zero.

**How did you choose alpha and gamma in the following equation?**

For baseline model, default values of alpha and gamma has been used and for tuned model those values of alpha and gamma are used from which agent received maximum Score over time.

**Try a policy other than maxQ(s', a'). How did it change the baseline performance?** 

Q- learning method is a value based approch so need to change the policy for this algorithm.

**What is the average number of steps taken per episode?**

0.37373737373737376

**Does Q-learning use value-based or policy-based iteration?**

Q-learning method used value-based iteration

## **Conclusion**

Score over time for baseline model is 0.1158 and after we tuned the hyperparameters Score over time of model became 0.145 which is approximately 28% increse in baseline model.
By changing value of epsilon and decay rate Score over time and average steps per episode increases. In this model we have decreased exploration tradeoff and. 




##**Contributions**

Referred the algorithm from citations and implemented by using reinforcement learning and Open AI to play Froze lake.

##**Citation**


1.   [Q-Learning with numpy](https://youtu.be/q2ZOEFAaaI0 ) 
2.   [Implementaion code](https://github.com/openai/gym)



##**License**

Copyright 2020 Yashraj Swarnkar

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.